lib/core/src/Cardano/Wallet/Network/Light.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+module Cardano.Wallet.Network.Light
+    ( -- * Interface
+      LightSyncSource (..)
+    , LightBlocks
+    , hoistLightSyncSource
+    , lightSync
+
+    , LightLayerLog (..)
+    ) where
+
+import Prelude
+
+import Cardano.BM.Data.Severity
+    ( Severity (..) )
+import Cardano.BM.Data.Tracer
+    ( HasPrivacyAnnotation (..), HasSeverityAnnotation (..) )
+import Cardano.Wallet.Network
+    ( ChainFollower (..) )
+import Cardano.Wallet.Primitive.BlockSummary
+    ( BlockSummary (..) )
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..)
+    , ChainPoint (..)
+    , chainPointFromBlockHeader
+    , compareSlot
+    )
+import Control.Monad.Class.MonadTimer
+    ( DiffTime, MonadDelay (..) )
+import Control.Tracer
+    ( Tracer, traceWith )
+import Data.List
+    ( maximumBy, sortBy )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Text.Class
+    ( ToText (..) )
+import Data.Void
+    ( Void )
+import Data.Word
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text as T
+
+{-------------------------------------------------------------------------------
+    LightLayer
+-------------------------------------------------------------------------------}
+type BlockHeight = Integer
+
+-- | Blockchain data source suitable for the implementation of 'lightSync'.
+data LightSyncSource m block addr txs = LightSyncSource
+    { stabilityWindow :: BlockHeight
+        -- ^ Stability window.
+    , getHeader :: block -> BlockHeader
+        -- ^ Get the 'BlockHeader' of a given @block@.
+    , getTip :: m BlockHeader
+        -- ^ Latest tip of the chain.
+    , isConsensus :: ChainPoint -> m Bool
+        -- ^ Check whether a 'ChainPoint' still exists in the consensus,
+        -- or whether the chain has rolled back already.
+    , getBlockHeaderAtHeight :: BlockHeight -> m (Maybe BlockHeader)
+        -- ^ Get the 'BlockHeader' at a given block height.
+        -- Returns 'Nothing' if there is no block at this height (anymore).
```

Comment:
Also thinking about source abstraction. Maybe it would be worth to sketch the abtraction that would handle this light wallet source and node source? I think we can do it later, now we can demonstrate it just works. But maybe we can do something now. What do you think? Too soon? Or you see already silluette of this abstraction?

Reply:
> Also thinking about source abstraction. 

I've had some discussion with Yura about this, and the conclusion was "too soon". I was thinking of some sort of general `LightLayer` abstraction indeed, but the details seemed unclear. We decided to postpone a decision there, and I choose to introduce the smallest type that can be used to implement `lightSync` instead.

Reply:
> not better to return whatever is present up to block height?

In the lightSync procedure, `getBlockHeaderAtHeight` is only used to move a pointer around, here to a block slightly older than the tip of the chain (by subtracting `stabilityWindow` from the height).

The actual data for blocks is returned by `getNextBlocks` and `getAddressTxs`.